internal/pkg/otel/core_components_test.go

```diff
@@ -0,0 +1,139 @@
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package otel
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"gopkg.in/yaml.v3"
+)
+
+// TestCoreComponentsInGoMod verifies that all components listed in core-components.yaml
+// are present in the project dependencies.
+func TestCoreComponentsInGoMod(t *testing.T) {
+	// Extract components from project dependencies
+	moduleComponents, err := extractComponentsFromDeps()
+	if err != nil {
+		t.Fatalf("Failed to extract components from dependencies: %v", err)
+	}
+
+	// Print found components for debugging
+	t.Logf("Found components in dependencies: %v", moduleComponents)
+
+	// Load components from core-components.yaml
+	yamlComponents, err := loadCoreComponentsYAML()
+	if err != nil {
+		t.Fatalf("Failed to load core-components.yaml: %v", err)
+	}
+
+	// Verify all components in YAML are present in dependencies
+	for _, component := range yamlComponents {
+		assert.Contains(t, moduleComponents, component)
+	}
+}
+
+// GoListModule represents the relevant parts of the `go list -json` output
+type GoListModule struct {
+	Deps []string `json:"Deps"`
+}
+
+// extractComponentsFromDeps extracts component names from the project dependencies
+// by running `go list -json` and parsing the output
+func extractComponentsFromDeps() ([]string, error) {
+	// Run go list -json to get the dependencies
+	cmd := exec.Command("go", "list", "-json", "github.com/elastic/elastic-agent")
+	output, err := cmd.Output()
+	if err != nil {
+		return nil, fmt.Errorf("failed to run go list command: %w", err)
+	}
+
+	// Parse the JSON output
+	var module GoListModule
+	if err := json.Unmarshal(output, &module); err != nil {
+		return nil, fmt.Errorf("failed to parse go list output: %w", err)
+	}
+
+	// Component types to look for in dependency paths
+	componentTypes := []string{
+		"processor",
+		"receiver",
+		"extension",
+		"exporter",
+		"connector",
+		"confmap/provider",
+		"storage",
+	}
+
+	// Extract component names from dependency paths
+	var components []string
+	for _, dep := range module.Deps {
+		for _, cType := range componentTypes {
+			// Check if the dependency path contains the component type
+			if strings.Contains(dep, "/"+cType+"/") {
+				// Extract the component name (last part of the path)
+				parts := strings.Split(dep, "/")
+				if len(parts) > 0 {
+					componentName := parts[len(parts)-1]
+					components = append(components, componentName)
+				}
+			}
+		}
+
+		// Special case for filestorage which is under extension/storage/
+		if strings.Contains(dep, "/extension/storage/") {
+			parts := strings.Split(dep, "/")
+			if len(parts) > 0 {
+				componentName := parts[len(parts)-1]
+				components = append(components, componentName)
+			}
+		}
```

Comment:
Not sure why `/extension/storage/` needs a specific treatment here while `confmap/provider` doesn't :thinking: